File: libembosser-core-drivers/src/main/java/org/brailleblaster/libembosser/drivers/utils/BaseGraphicsEmbosser.java
package org.brailleblaster.libembosser.drivers.utils;

import java.awt.print.PageFormat;
import java.awt.print.Paper;
import java.awt.print.Printable;
import java.awt.print.PrinterException;
import java.awt.print.PrinterJob;
import java.io.InputStream;
import java.util.Optional;

import javax.print.DocFlavor;
import javax.print.PrintService;
import javax.print.StreamPrintServiceFactory;
import javax.print.attribute.HashPrintRequestAttributeSet;
import javax.print.attribute.PrintRequestAttribute;
import javax.print.attribute.PrintRequestAttributeSet;
import javax.print.attribute.standard.Sides;

import org.brailleblaster.libembosser.drivers.utils.DocumentParser.ParseException;
import org.brailleblaster.libembosser.drivers.utils.DocumentToPrintableHandler.LayoutHelper;
import org.brailleblaster.libembosser.embossing.attribute.Copies;
import org.brailleblaster.libembosser.embossing.attribute.PageRanges;
import org.brailleblaster.libembosser.embossing.attribute.PaperLayout;
import org.brailleblaster.libembosser.embossing.attribute.PaperMargins;
import org.brailleblaster.libembosser.spi.BrlCell;
import org.brailleblaster.libembosser.spi.EmbossException;
import org.brailleblaster.libembosser.spi.Embosser;
import org.brailleblaster.libembosser.spi.EmbossingAttributeSet;
import org.w3c.dom.Document;

/**
 * Base class for embossers using Java2D graphics.
 * 
 * @author Michael Whapples
 *
 */
public abstract class BaseGraphicsEmbosser implements Embosser {
	private StreamPrintServiceFactory[] streamPrintServiceFactories = PrinterJob.lookupStreamPrintServices(DocFlavor.BYTE_ARRAY.POSTSCRIPT.getMimeType());
	private String id;
	private String manufacturer;
	private String model;
	
	protected BaseGraphicsEmbosser(String id, String manufacturer, String model) {
		this.id = id;
		this.manufacturer = manufacturer;
		this.model = model;
	}
	
	/**
	 * Get a suitable font for the Braille cell type.
	 * 
	 * @param cell The cell type to be embossed.
	 * @return A suitable font for printing the Braille cell type.
	 */
	public abstract LayoutHelper getLayoutHelper(BrlCell cell);

	@Override
	public void embossPef(PrintService embosserDevice, Document pef, EmbossingAttributeSet attributes)
			throws EmbossException {
		emboss(embosserDevice, pef, attributes, new DocumentParser()::parsePef);
	}

	@Override
	public void embossPef(PrintService embosserDevice, InputStream pef, EmbossingAttributeSet attributes)
			throws EmbossException {
		emboss(embosserDevice, pef, attributes, new DocumentParser()::parsePef);
	}

	@Override
	public void embossBrf(PrintService embosserDevice, InputStream brf, EmbossingAttributeSet attributes)
			throws EmbossException {
		emboss(embosserDevice, brf, attributes, new DocumentParser()::parseBrf);
	}
	private <T> void emboss(PrintService ps, T input, EmbossingAttributeSet attributes, ThrowingBiConsumer<T, DocumentHandler, ParseException> parseMethod) throws EmbossException {
		PageRanges pages = Optional.ofNullable((PageRanges)attributes.get(PageRanges.class)).orElseGet(() -> new PageRanges());
		PageFilterHandler<DocumentToPrintableHandler> pageFilteredHandler = new PageFilterHandler<DocumentToPrintableHandler>(new DocumentToPrintableHandler.Builder().setLayoutHelper(getLayoutHelper(BrlCell.NLS)).build(), pages);
		try {
			parseMethod.accept(input, pageFilteredHandler);
		} catch (ParseException e) {
			throw new RuntimeException("Problem parsing document", e);
		}
		DocumentToPrintableHandler handler = pageFilteredHandler.getDelegate();
		Printable printable = handler.asPrintable();
		PrinterJob printJob = PrinterJob.getPrinterJob();
		printJob.setJobName("BrailleBlasterEmboss");
		Optional.ofNullable(attributes.get(Copies.class)).map(v -> ((org.brailleblaster.libembosser.embossing.attribute.Copies)v).getValue()).ifPresent(v -> printJob.setCopies(v));
		PrintRequestAttribute duplex = Optional.ofNullable((org.brailleblaster.libembosser.embossing.attribute.PaperLayout)attributes.get(PaperLayout.class)).filter(p -> supportsInterpoint()).map(p -> p.getValue().isDoubleSide() ? Sides.DUPLEX : Sides.ONE_SIDED).orElse(Sides.ONE_SIDED);
		try {
			printJob.setPrintService(ps);
			PageFormat pf = printJob.defaultPage();
			Paper paper = pf.getPaper();
			Optional.ofNullable(attributes.get(org.brailleblaster.libembosser.embossing.attribute.PaperSize.class)).map(p -> ((org.brailleblaster.libembosser.embossing.attribute.PaperSize)p).getValue()).ifPresent(r -> paper.setSize(mmToPt(r.getWidth().doubleValue()), mmToPt(r.getHeight().doubleValue())));
			final double width = paper.getWidth();
			final double height = paper.getHeight();
			Optional.ofNullable(attributes.get(org.brailleblaster.libembosser.embossing.attribute.PaperMargins.class)).map(m -> ((PaperMargins)m).getValue()).ifPresent(m -> {
				final double left = mmToPt(m.getLeft().doubleValue());
				// final double right = mmToPt(m.getRight().doubleValue());
				final double top = mmToPt(m.getTop().doubleValue());
				// final double bottom = mmToPt(m.getBottom().doubleValue());
				paper.setImageableArea(left, top, width - left, height - top);
			});
			pf.setPaper(paper);
			printJob.setPrintable(printable, pf);
			PrintRequestAttributeSet requestAttributes = new HashPrintRequestAttributeSet(duplex);
			printJob.print(requestAttributes);
		} catch (PrinterException e) {
			throw new EmbossException("Problem sending emboss job to embosser device.", e);
		}
	}
	
	private double mmToPt(double mm) {
		return (mm * 72.0)/25.4;
	}
	
	@Override
	public Optional<StreamPrintServiceFactory> getStreamPrintServiceFactory() {
		return streamPrintServiceFactories.length > 0? Optional.of(streamPrintServiceFactories[0]) : Optional.empty();
	}
	
	@Override
	public String getId() {
		// TODO Auto-generated method stub
		return id;
	}
	
	@Override
	public String getManufacturer() {
		// TODO Auto-generated method stub
		return manufacturer;
	}
	
	@Override
	public String getModel() {
		// TODO Auto-generated method stub
		return model;
	}
	
}

File: libembosser-core-drivers/src/main/kotlin/org/brailleblaster/libembosser/drivers/utils/BaseGraphicsEmbosser.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * Copyright (c) 2023 American Printing House for the Blind
 */

package org.brailleblaster.libembosser.drivers.utils

import com.google.common.collect.ImmutableSet
import org.brailleblaster.libembosser.drivers.utils.DocumentToPrintableHandler.LayoutHelper
import org.brailleblaster.libembosser.drivers.utils.document.events.DocumentEvent
import org.brailleblaster.libembosser.drivers.utils.document.filters.InterpointGraphicTransform
import org.brailleblaster.libembosser.drivers.utils.document.filters.PageFilter
import org.brailleblaster.libembosser.embossing.attribute.*
import org.brailleblaster.libembosser.embossing.attribute.PaperSize
import org.brailleblaster.libembosser.spi.*
import org.w3c.dom.Document
import java.awt.print.PrinterException
import java.awt.print.PrinterJob
import java.io.InputStream
import java.util.*
import java.util.function.Function
import javax.print.DocFlavor
import javax.print.PrintService
import javax.print.StreamPrintServiceFactory
import javax.print.attribute.Attribute
import javax.print.attribute.HashPrintRequestAttributeSet
import javax.print.attribute.PrintRequestAttribute
import javax.print.attribute.PrintRequestAttributeSet
import javax.print.attribute.standard.Sides

/**
 * Base class for embossers using Java2D graphics.
 *
 * @author Michael Whapples
 */
abstract class BaseGraphicsEmbosser protected constructor(private val id: String, private val manufacturer: String, private val model: String) : Embosser {
    private val streamPrintServiceFactories = PrinterJob.lookupStreamPrintServices(DocFlavor.BYTE_ARRAY.POSTSCRIPT.mimeType)

    /**
     * Get a suitable font for the Braille cell type.
     *
     * @param cell The cell type to be embossed.
     * @return A suitable font for printing the Braille cell type.
     */
    abstract fun getLayoutHelper(cell: BrlCell?): LayoutHelper
    @Throws(EmbossException::class)
    override fun embossPef(embosserDevice: PrintService, pef: Document, attributes: EmbossingAttributeSet) {
        emboss(embosserDevice, pef, attributes) { inputDoc: Document, handler: DocumentHandler ->
            DocumentParser().parsePef(
                inputDoc,
                handler
            )
        }
    }

    @Throws(EmbossException::class)
    override fun embossBrf(embosserDevice: PrintService, brf: InputStream, attributes: EmbossingAttributeSet) {
        emboss(embosserDevice, brf, attributes) { input: InputStream, handler: DocumentHandler ->
            DocumentParser().parseBrf(
                input,
                handler
            )
        }
    }

    @Throws(EmbossException::class)
    private fun <T> emboss(ps: PrintService, input: T, attributes: EmbossingAttributeSet, parseMethod: ThrowingBiConsumer<T, DocumentHandler, DocumentParser.ParseException>) {
        val events: MutableList<DocumentEvent> = LinkedList()
        try {
            parseMethod.accept(input, DocumentHandler { e: DocumentEvent -> events.add(e) })
        } catch (e: DocumentParser.ParseException) {
            throw RuntimeException("Problem parsing document", e)
        }
        val duplex: PrintRequestAttribute = Optional.ofNullable(attributes[PaperLayout::class.java] as PaperLayout).filter { supportsInterpoint() }.filter { p: PaperLayout -> p.value.isDoubleSide }.map { Sides.TWO_SIDED_LONG_EDGE }.orElse(Sides.ONE_SIDED)
        val pages = Optional.ofNullable(attributes[PageRanges::class.java] as PageRanges).orElseGet { PageRanges() }
        var transform: Function<Iterator<DocumentEvent?>?, Iterator<DocumentEvent?>?> = PageFilter(pages)
        if (DOUBLE_SIDED_MODES.contains(duplex)) {
            transform = transform.andThen(InterpointGraphicTransform())
        }
        val handler = transform.andThen(DocumentToPrintableHandler.Builder().setLayoutHelper(getLayoutHelper(BrlCell.NLS)).build())
        val printable = handler.apply(events.iterator())
        val printJob = PrinterJob.getPrinterJob()
        printJob.jobName = "BrailleBlasterEmboss"
        Optional.ofNullable(attributes[Copies::class.java]).map { v: Attribute -> (v as Copies).value }.ifPresent { copies -> printJob.copies = copies }
        try {
            printJob.printService = ps
            val pf = printJob.defaultPage()
            val paper = pf.paper
            Optional.ofNullable(attributes[PaperSize::class.java]).map { p: Attribute -> (p as PaperSize).value }.ifPresent { r: Rectangle -> paper.setSize(mmToPt(r.width.toDouble()), mmToPt(r.height.toDouble())) }
            val width = paper.width
            val height = paper.height
            Optional.ofNullable(attributes[PaperMargins::class.java]).map { m: Attribute -> (m as PaperMargins).value }.ifPresent { m: Margins ->
                val left = mmToPt(m.left.toDouble())
                // final double right = mmToPt(m.getRight().doubleValue());
                val top = mmToPt(m.top.toDouble())
                // final double bottom = mmToPt(m.getBottom().doubleValue());
                paper.setImageableArea(left, top, width - left, height - top)
            }
            pf.paper = paper
            printJob.setPrintable(printable, pf)
            val requestAttributes: PrintRequestAttributeSet = HashPrintRequestAttributeSet(duplex)
            printJob.print(requestAttributes)
        } catch (e: PrinterException) {
            throw EmbossException("Problem sending emboss job to embosser device.", e)
        }
    }

    private fun mmToPt(mm: Double): Double {
        return mm * 72.0 / 25.4
    }

    override fun getStreamPrintServiceFactory(): Optional<StreamPrintServiceFactory> {
        return if (streamPrintServiceFactories.isNotEmpty()) Optional.of(streamPrintServiceFactories[0]) else Optional.empty()
    }

    override fun getId(): String {
        // TODO Auto-generated method stub
        return id
    }

    override fun getManufacturer(): String {
        // TODO Auto-generated method stub
        return manufacturer
    }

    override fun getModel(): String {
        // TODO Auto-generated method stub
        return model
    }

    companion object {
        val DOUBLE_SIDED_MODES: Set<Sides> = ImmutableSet.of(Sides.DUPLEX, Sides.TWO_SIDED_LONG_EDGE, Sides.TUMBLE, Sides.TWO_SIDED_SHORT_EDGE)
    }
}